jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/PathResourceTest.java

```diff
@@ -140,4 +158,204 @@ public void testSame()
+                supportsUtf8Dir = true;
+            }
+            catch (InvalidPathException e)
+            {
+                LOG.debug("IGNORE", e);
+                supportsUtf8Dir = false;
+            }
+        }
+
+        try (ResourceFactory.Closeable resourceFactory = ResourceFactory.closeable())
+        {
+            Resource archiveResource = resourceFactory.newResource(jarUri);
+
+            Resource testText = archiveResource.resolve("/dir/test.txt");
+            assertTrue(testText.exists());
+            assertFalse(testText.isAlias());
+
+            // Resolve file to name, but different case
+            testText = archiveResource.resolve("/dir/TEST.TXT");
+            assertFalse(testText.exists());
+            testText = archiveResource.resolve("/DIR/test.txt");
+            assertFalse(testText.exists());
+
+            // Resolve file using path navigation
+            testText = archiveResource.resolve("/foo/../dir/test.txt");
+            assertTrue(testText.exists());
+            assertFalse(testText.isAlias());
+
+            // Resolve file using encoded characters
+            testText = archiveResource.resolve("/dir/test%2Etxt");
+            assertTrue(testText.exists());
+            assertFalse(testText.isAlias());
+
+            // Resolve file using extension-less directory
+            testText = archiveResource.resolve("/dir./test.txt");
+            assertFalse(testText.exists());
+            assertFalse(testText.isAlias());
+
+            // Resolve directory to name, no slash
+            Resource dirResource = archiveResource.resolve("/dir");
+            assertTrue(dirResource.exists());
+            assertFalse(dirResource.isAlias());
+
+            // Resolve directory to name, with slash
+            dirResource = archiveResource.resolve("/dir/");
+            assertTrue(dirResource.exists());
+            assertFalse(dirResource.isAlias());
+
+            if (supportsUtf8Dir)
+            {
+                // Resolve utf8 directory in raw
+                dirResource = archiveResource.resolve("/bãm/");
+                assertTrue(dirResource.exists());
+                assertFalse(dirResource.isAlias());
+
+                // Resolve utf8 directory encoded
+                dirResource = archiveResource.resolve("/b%C3%A3m/");
+                assertTrue(dirResource.exists());
+                assertFalse(dirResource.isAlias());
+            }
+        }
+    }
+
+    @Test
+    public void testNullCharEndingFilename(WorkDir workDir) throws Exception
+    {
+        Path testJar = workDir.getEmptyPathDir().resolve("test.jar");
+
+        Map<String, String> env = new HashMap<>();
+        env.put("create", "true");
+
+        URI jarUri = URIUtil.uriJarPrefix(testJar.toUri(), "!/");
+        try (FileSystem zipfs = FileSystems.newFileSystem(jarUri, env))
+        {
+            Path root = zipfs.getPath("/");
+
+            Path file0 = root.resolve("test.txt\0");
```

Comment:
On ZipFS, this `Path.resolve("test.txt\0")` results in a `InvalidPathException`...

```
java.nio.file.InvalidPathException: Path: nul character not allowed: test.txt?
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.normalize(ZipPath.java:501)
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.normalize(ZipPath.java:474)
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.normalize(ZipPath.java:516)
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.resolve(ZipPath.java:370)
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.resolve(ZipPath.java:56)
	at org.eclipse.jetty.util@12.0.0-SNAPSHOT/org.eclipse.jetty.util.resource.PathResourceTest.testNullCharEndingFilename(PathResourceTest.java:299)
```